### PR TITLE
[SPARK-17720][SQL] introduce static SQL conf

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

SQLConf is session-scoped and mutable. However, we do have the requirement for a static SQL conf, which is global and immutable, e.g. the `schemaStringThreshold` in `HiveExternalCatalog`, the flag to enable/disable hive support, the global temp view database in https://github.com/apache/spark/pull/14897.

Actually we've already implemented static SQL conf implicitly via `SparkConf`, this PR just make it explicit and expose it to users, so that they can see the config value via SQL command or `SparkSession.conf`, and forbid users to set/unset static SQL conf.
## How was this patch tested?

new tests in SQLConfSuite

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai @gatorsmile @liancheng

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 1]
```diff
@@ -791,7 +791,7 @@ object SparkSession {
       // Get the session from current thread's active session.
       var session = activeThreadSession.get()
       if ((session ne null) && !session.sparkContext.isStopped) {
-        options.foreach { case (k, v) => session.conf.set(k, v) }
+        options.foreach { case (k, v) => session.sessionState.conf.setConfString(k, v) }
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: this is change related?

**Comment by andrewor14** [comment]
LGTM. Pretty straightforward.

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: yea, the goal is to forbid users to set/unset global sql conf via public API, but here we do need to set global sql conf with sparkConf, so we should use internal API here instead.

**Comment by andrewor14** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala [hunk 2]
```diff
@@ -250,4 +253,20 @@ class SQLConfSuite extends QueryTest with SharedSQLContext {
...
+
+    assert(newSession.conf.get(SCHEMA_STRING_LENGTH_THRESHOLD.key) == "2000")
+    checkAnswer(
+      newSession.sql(s"SET ${SCHEMA_STRING_LENGTH_THRESHOLD.key}"),
+      Row(SCHEMA_STRING_LENGTH_THRESHOLD.key, "2000"))
+  }
+
+  test("cannot set/unset global SQL conf") {
+    intercept[AnalysisException](sql(s"SET ${SCHEMA_STRING_LENGTH_THRESHOLD.key}=10"))
+    intercept[AnalysisException](sql(s"UNSET ${SCHEMA_STRING_LENGTH_THRESHOLD.key}=10"))
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: We do not support UNSET command in SQL interface. 

Maybe you can change it to 
```
spark.conf.unset(SCHEMA_STRING_LENGTH_THRESHOLD.key)
```

File: sql/core/src/main/scala/org/apache/spark/sql/RuntimeConfig.scala [hunk 3]
```diff
@@ -36,6 +37,7 @@ class RuntimeConfig private[sql](sqlConf: SQLConf = new SQLConf) {
    * @since 2.0.0
    */
   def set(key: String, value: String): Unit = {
+    assertNotGlobalSQLConf(key)
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/RuntimeConfig.scala]: I am wondering if we need to add this `assertNotGlobalSQLConf ` in all the public set functions, right? I am just afraid we might forget to update the other `set` functions when adding more global sqlconf.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -877,3 +861,36 @@ private[sql] class SQLConf extends Serializable with CatalystConf with Logging {
...
+    }
+  }
+
+  val CATALOG_IMPLEMENTATION = buildConf("spark.sql.catalogImplementation")
+    .internal()
+    .stringConf
+    .checkValues(Set("hive", "in-memory"))
+    .createWithDefault("in-memory")
+
+  // This is used to control the when we will split a schema's JSON string to multiple pieces
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Nit: remove `the`

File: sql/core/src/main/scala/org/apache/spark/sql/RuntimeConfig.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/RuntimeConfig.scala]: all global sql conf will be automatically registered, we don't need to change the code here when adding more global sql conf

File: sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: oh damn, but why test passed... it should throw parser exception

**Comment by rxin** [comment]
"Global" isn't the right name. The main property is immutability once the service starts, rather than the fact that they are shared by all the sessions.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala [hunk 5]
```diff
@@ -699,28 +699,27 @@ class MetastoreDataSourcesSuite extends QueryTest with SQLTestUtils with TestHiv
   }
 
   test("SPARK-6024 wide schema support") {
-    withSQLConf(SQLConf.SCHEMA_STRING_LENGTH_THRESHOLD.key -> "4000") {
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/MetastoreDataSourcesSuite.scala]: This line in the old test case is useless... : (

File: sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/internal/SQLConfSuite.scala]: `ParseException` extends `AnalysisException`

File: sql/core/src/main/scala/org/apache/spark/sql/RuntimeConfig.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/RuntimeConfig.scala]: Sorry, I did not explain it clearly. What I meant above is to add `assertNotGlobalSQLConf` to  [the other two set functions](https://github.com/cloud-fan/spark/blob/59fc4e996c60359e3d1b232fa3699c98a3f27f9e/sql/core/src/main/scala/org/apache/spark/sql/RuntimeConfig.scala#L44-L60) Does this make sense?

**Comment by rxin** [comment]
@srinathshankar suggested "static config"

File: sql/core/src/main/scala/org/apache/spark/sql/RuntimeConfig.scala [hunk 6]
```diff
@@ -132,4 +135,9 @@ class RuntimeConfig private[sql](sqlConf: SQLConf = new SQLConf) {
     sqlConf.contains(key)
   }
 
+  private def assertNotGlobalSQLConf(key: String): Unit = {
+    if (GlobalSQLConf.globalConfKeys.contains(key)) {
+      throw new AnalysisException(s"Can not set/unset a global SQL conf: $key")
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/RuntimeConfig.scala]: Cannot modify the value of a static config: $key

File: sql/core/src/main/scala/org/apache/spark/sql/RuntimeConfig.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/RuntimeConfig.scala]: ok, looks safer

**Comment by felixcheung** [comment]
Should we need to update https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala#L67
?

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/RuntimeConfig.scala (hunk 3 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/RuntimeConfig.scala]: should be assertNonStaticConf here

**Comment by rxin** [comment]
LGTM except that one comment on naming.

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Merging to master! Thanks!

**Pull request closed without merging**